In [6]:
import numpy as np

# Условия выигрышей
winnings = {
    "777": 200,
    "999": 100,
    "555": 50,
    "333": 15,
    "111": 10,
    "*77": 5,
    "**7": 3,
    "*00": 2,
    "**0": 1,
}

# Количество испытаний
trials = 1_000_000
total_profit = 0  # Общая прибыль
successes = {key: 0 for key in winnings}  # Частота выигрышей по каждому правилу

# Моделирование игры () Используем метод Монте-Карло для моделирования 1,000,000 игр

for _ in range(trials):
    number = np.random.randint(0, 1000)  # Случайное число от 0 до 999
    number_str = str(number).zfill(3)  # Преобразуем в строку с ведущими нулями

    # Определяем выигрыш
    if number_str == "777":
        successes["777"] += 1
        total_profit += 200
    elif number_str == "999":
        successes["999"] += 1
        total_profit += 100
    elif number_str == "555":
        successes["555"] += 1
        total_profit += 50
    elif number_str == "333":
        successes["333"] += 1
        total_profit += 15
    elif number_str == "111":
        successes["111"] += 1
        total_profit += 10
    elif number_str.endswith("77"):
        successes["*77"] += 1
        total_profit += 5
    elif number_str.endswith("7"):
        successes["**7"] += 1
        total_profit += 3
    elif number_str.endswith("00"):
        successes["*00"] += 1
        total_profit += 2
    elif number_str.endswith("0"):
        successes["**0"] += 1
        total_profit += 1

    # Учитываем стоимость игры
    total_profit -= 1

# Подсчет результатов
average_profit = total_profit / trials
success_probabilities = {key: count / trials for key, count in successes.items()}

print(f"Средняя прибыль за игру: {average_profit:.2f} руб.")
print("Частоты выигрышей (приближенные вероятности):")
for rule, prob in success_probabilities.items():
    print(f"  {rule}: {prob:.6f}")


Средняя прибыль за игру: -0.21 руб.
Частоты выигрышей (приближенные вероятности):
  777: 0.000981
  999: 0.000967
  555: 0.000987
  333: 0.001050
  111: 0.000989
  *77: 0.008977
  **7: 0.090094
  *00: 0.010076
  **0: 0.089551


Что делает код:
Случайное число (0–999): Генерируется с помощью np.random.randint.
Проверка условий выигрыша: Каждое число проверяется на соответствие условиям из задачи.
Подсчет частоты выигрышей: Частота каждого события обновляется.
Средняя прибыль: Общая прибыль делится на количество игр, чтобы получить среднюю прибыль за одну игру.


Ожидаемые результаты:
Средняя прибыль за игру будет отрицательной, так как игра явно выгодна для автомата, а не для игрока.
Частоты выигрышей будут близки к теоретическим вероятностям при большом числе испытаний.

Средняя прибыль за игру
Средняя прибыль/убыток за игру: 

−0.21руб.

Это говорит о том, что игрок теряет в среднем 21 копейку за каждую игру. Следовательно, с точки зрения игрока игра невыгодна. Для организатора же эта игра является прибыльной на длинной дистанци

Частоты выигрышей (приближенные вероятности)
Частоты указывают, как часто выпадает каждое из выигрышных условий. Вот что это значит:

Выигрыш	Частота	Интерпретация

777	0.000969	Выпадает примерно 1 раз на 1000 игр.

999	0.000994	Аналогично, редкий случай.

555	0.001043	Чуть более вероятно, но всё ещё редко.

333	0.000964	Очень малая вероятность.

111	0.000954	Также малая вероятность.

*77	0.009018	Чаще, примерно 9 раз на 1000 игр.

**7	0.089666	Вероятность около 9%, достаточно часто.

*00	0.009991	Почти 1%.

**0	0.090154	Примерно 9%.